dll/win32/shell32/CShellDispatch.cpp

```diff
@@ -281,19 +319,106 @@ HRESULT STDMETHODCALLTYPE CShellDispatch::FindPrinter(BSTR name, BSTR location,
 HRESULT STDMETHODCALLTYPE CShellDispatch::GetSystemInformation(BSTR name, VARIANT *ret)
 {
     TRACE("(%p, %ls, %p)\n", this, name, ret);
+
+    if (!lstrcmpiW(name, L"ProcessorArchitecture"))
+    {
+        SYSTEM_INFO si;
+        GetSystemInfo(&si);
+        V_VT(ret) = VT_I4;
+        V_UI4(ret) = si.wProcessorArchitecture;
+        return S_OK;
+    }
+
+    UINT os = 0;
+    if (!lstrcmpiW(name, L"IsOS_Professional"))
+        os = OS_PROFESSIONAL;
+    else if (!lstrcmpiW(name, L"IsOS_Personal"))
+        os = OS_HOME;
+    else if (!lstrcmpiW(name, L"IsOS_DomainMember"))
+        os = OS_DOMAINMEMBER;
+    if (os)
+    {
+        V_VT(ret) = VT_BOOL;
+        V_BOOL(ret) = IsOS(os) ? VARIANT_TRUE : VARIANT_FALSE;
+        return S_OK;
+    }
+
     return E_NOTIMPL;
 }
 
+static SC_HANDLE OpenServiceHelper(LPCWSTR name, DWORD access)
+{
+    SC_HANDLE hSvc = NULL, hScm = OpenSCManagerW(NULL, NULL, SC_MANAGER_CONNECT);
+    if (hScm)
+    {
+        hSvc = OpenServiceW(hScm, name, access);
+        CloseServiceHandle(hScm);
+    }
+    return hSvc;
+}
+
+static HRESULT SHELL32_ControlService(BSTR name, DWORD control, VARIANT &persistent)
+{
+    BOOL persist = V_VT(&persistent) == VT_BOOL && V_BOOL(&persistent);
+    DWORD access = persist ? SERVICE_CHANGE_CONFIG : 0;
+    switch(control)
+    {
+        case 0:
+            access |= SERVICE_START;
+            break;
+        case SERVICE_CONTROL_STOP:
+            access |= SERVICE_STOP;
+            break;
+    }
+    HRESULT hr = HRESULT_FROM_WIN32(ERROR_ACCESS_DENIED);
+    SC_HANDLE hSvc = OpenServiceHelper(name, access);
+    if (hSvc)
+    {
```

Comment:
Looks like you could simplify as follows:
```suggestion
    if (!hSvc)
        return HRESULT_FROM_WIN32(ERROR_ACCESS_DENIED);

    BOOL success;
// etc. as below, with the indentation level down.
```
Note that the `hr` variable would then be set in any case via the `hr = success ? ...` line.

Regarding the `HRESULT_FROM_WIN32(ERROR_ACCESS_DENIED)`: isn't there a slightly more accurate error? If the OpenServiceHelper fails, I'm not sure this is always due to a denied access.